File: sub_pub/adapters/mock.py
"""Mock adapter for testing and development"""
import asyncio
import logging
import time
from typing import AsyncIterator, Iterator, List, Optional
from datetime import datetime

from sub_pub.core.interfaces import (
    AsyncMessagePublisher,
    AsyncMessageSource,
    MessagePublisher,
    MessageSource,
)
from sub_pub.core.message import Message

logger = logging.getLogger(__name__)


class MockSource(MessageSource):
    """Mock message source for testing"""
    
    def __init__(self, config: dict):
        self.config = config
        self.subscribed_topics: List[str] = []
        self._connected = False
        self._message_count = 0
        
    def connect(self) -> None:
        """Establish connection"""
        logger.info("Mock source connected")
        self._connected = True
        
    def subscribe(self, topics: List[str]) -> None:
        """Subscribe to topics"""
        self.subscribed_topics = topics
        logger.info(f"Mock source subscribed to: {topics}")
        
    def consume(self) -> Iterator[Message]:
        """Consume messages (generates mock messages)"""
        while self._connected:
            for topic in self.subscribed_topics:
                self._message_count += 1
                yield Message(
                    payload=f"Mock message {self._message_count}".encode('utf-8'),
                    headers={'source': 'mock', 'count': str(self._message_count)},
                    topic=topic,
                    key=f"key-{self._message_count}",
                    timestamp=datetime.now()
                )
                time.sleep(0.1)  # Slow down for testing
                
    def close(self) -> None:
        """Close connection"""
        self._connected = False
        logger.info("Mock source closed")
        
    def commit(self, message: Optional[Message] = None) -> None:
        """Commit message offset"""
        pass  # No-op for mock


class MockPublisher(MessagePublisher):
    """Mock message publisher for testing"""
    
    def __init__(self, config: dict):
        self.config = config
        self._connected = False
        self.published_messages: List[tuple[Message, str]] = []
        
    def connect(self) -> None:
        """Establish connection"""
        logger.info("Mock publisher connected")
        self._connected = True
        
    def publish(self, message: Message, topic: str) -> None:
        """Publish a message"""
        self.published_messages.append((message, topic))
        logger.debug(f"Mock published to {topic}: {len(message.payload)} bytes")
        
    def flush(self) -> None:
        """Flush buffered messages"""
        logger.info(f"Mock publisher flushed {len(self.published_messages)} messages")
        
    def close(self) -> None:
        """Close connection"""
        self._connected = False
        logger.info("Mock publisher closed")


class AsyncMockSource(AsyncMessageSource):
    """Async mock message source for testing"""

    def __init__(self, messages: Optional[List[Message]] = None):
        self.subscribed_topics: List[str] = []
        self._connected = False
        self._message_count = 0
        # Pre-defined messages to yield (if None, generates mock messages)
        self._messages = messages

    async def connect(self) -> None:
        """Establish connection"""
        logger.info("Async mock source connected")
        self._connected = True

    async def subscribe(self, topics: List[str]) -> None:
        """Subscribe to topics"""
        self.subscribed_topics = topics
        logger.info(f"Async mock source subscribed to: {topics}")

    async def consume(self) -> AsyncIterator[Message]:
        """Consume messages (yields pre-defined messages or generates mock ones)"""
        if self._messages is not None:
            for message in self._messages:
                if not self._connected:
                    break
                yield message
        else:
            while self._connected:
                for topic in self.subscribed_topics:
                    if not self._connected:
                        return
                    self._message_count += 1
                    yield Message(
                        payload=f"Async mock message {self._message_count}".encode("utf-8"),
                        headers={"source": "async_mock", "count": str(self._message_count)},
                        topic=topic,
                        key=f"key-{self._message_count}",
                        timestamp=datetime.now(),
                    )
                    await asyncio.sleep(0.1)

    async def close(self) -> None:
        """Close connection"""
        self._connected = False
        logger.info("Async mock source closed")

    async def commit(self, message: Optional[Message] = None) -> None:
        """Commit message offset"""
        pass  # No-op for mock


class AsyncMockPublisher(AsyncMessagePublisher):
    """Async mock message publisher for testing"""

    def __init__(self):
        self._connected = False
        self.published_messages: List[tuple[Message, str]] = []

    async def connect(self) -> None:
        """Establish connection"""
        logger.info("Async mock publisher connected")
        self._connected = True

    async def publish(self, message: Message, topic: str) -> None:
        """Publish a message"""
        self.published_messages.append((message, topic))
        logger.debug(f"Async mock published to {topic}: {len(message.payload)} bytes")

    async def flush(self) -> None:
        """Flush buffered messages"""
        logger.info(f"Async mock publisher flushed {len(self.published_messages)} messages")

    async def close(self) -> None:
        """Close connection"""
        self._connected = False
        logger.info("Async mock publisher closed")

File: rust/src/adapters/mock.rs
use std::collections::VecDeque;
use std::sync::{Arc, Mutex};
use std::time::Duration;

use async_trait::async_trait;
use tokio::time::sleep;

use crate::core::{Message, MessagePublisher, MessageSource};
use crate::error::SubPubError;

// ---------------------------------------------------------------------------
// MockSource
// ---------------------------------------------------------------------------

/// A test-only [`MessageSource`] that generates synthetic messages at a fixed
/// rate.  Useful for integration tests, demos, and local development without
/// a real message system.
pub struct MockSource {
    topics: Vec<String>,
    /// Counter that increments with each generated message.
    count: u64,
    /// How long to pause between messages (default: 100 ms).
    interval: Duration,
    /// Whether the source has been connected.
    connected: bool,
}

impl MockSource {
    pub fn new() -> Self {
        Self {
            topics: Vec::new(),
            count: 0,
            interval: Duration::from_millis(100),
            connected: false,
        }
    }

    /// Override the interval between generated messages.
    pub fn with_interval(mut self, interval: Duration) -> Self {
        self.interval = interval;
        self
    }
}

impl Default for MockSource {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl MessageSource for MockSource {
    async fn connect(&mut self) -> Result<(), SubPubError> {
        self.connected = true;
        tracing::debug!("MockSource connected");
        Ok(())
    }

    async fn subscribe(&mut self, topics: &[&str]) -> Result<(), SubPubError> {
        self.topics = topics.iter().map(|t| t.to_string()).collect();
        tracing::debug!(?self.topics, "MockSource subscribed");
        Ok(())
    }

    async fn consume(&mut self) -> Result<Option<Message>, SubPubError> {
        if !self.connected {
            return Err(SubPubError::Connection(
                "MockSource not connected".to_string(),
            ));
        }
        if self.topics.is_empty() {
            return Ok(None);
        }

        sleep(self.interval).await;

        self.count += 1;
        let topic = self.topics[((self.count - 1) as usize) % self.topics.len()].clone();
        let payload = format!("mock message #{}", self.count).into_bytes();

        let mut msg = Message::new(payload, topic);
        msg.headers
            .insert("source".to_string(), "mock".to_string());
        msg.headers
            .insert("count".to_string(), self.count.to_string());

        Ok(Some(msg))
    }

    async fn commit(&mut self, _message: &Message) -> Result<(), SubPubError> {
        Ok(())
    }

    async fn close(&mut self) -> Result<(), SubPubError> {
        self.connected = false;
        tracing::debug!("MockSource closed");
        Ok(())
    }
}

// ---------------------------------------------------------------------------
// MockPublisher
// ---------------------------------------------------------------------------

/// A test-only [`MessagePublisher`] that stores all published messages in
/// memory so tests can inspect them.
#[derive(Clone, Default)]
pub struct MockPublisher {
    /// Shared storage; clones of the publisher see the same messages.
    messages: Arc<Mutex<VecDeque<Message>>>,
    connected: bool,
}

impl MockPublisher {
    pub fn new() -> Self {
        Self::default()
    }

    /// Drain and return all messages published so far.
    pub fn drain(&self) -> Vec<Message> {
        self.messages
            .lock()
            .expect("mock publisher lock poisoned")
            .drain(..)
            .collect()
    }

    /// Return a snapshot of all published messages without draining.
    pub fn messages(&self) -> Vec<Message> {
        self.messages
            .lock()
            .expect("mock publisher lock poisoned")
            .iter()
            .cloned()
            .collect()
    }

    /// How many messages have been published so far.
    pub fn len(&self) -> usize {
        self.messages
            .lock()
            .expect("mock publisher lock poisoned")
            .len()
    }

    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
}

#[async_trait]
impl MessagePublisher for MockPublisher {
    async fn connect(&mut self) -> Result<(), SubPubError> {
        self.connected = true;
        tracing::debug!("MockPublisher connected");
        Ok(())
    }

    async fn publish(&mut self, message: &Message, topic: &str) -> Result<(), SubPubError> {
        if !self.connected {
            return Err(SubPubError::Connection(
                "MockPublisher not connected".to_string(),
            ));
        }
        let mut msg = message.clone();
        msg.topic = topic.to_string();
        self.messages
            .lock()
            .expect("mock publisher lock poisoned")
            .push_back(msg);
        Ok(())
    }

    async fn flush(&mut self) -> Result<(), SubPubError> {
        Ok(())
    }

    async fn close(&mut self) -> Result<(), SubPubError> {
        self.connected = false;
        tracing::debug!("MockPublisher closed");
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::time::Duration;

    #[tokio::test]
    async fn mock_publisher_stores_messages() {
        let mut pub_ = MockPublisher::new();
        pub_.connect().await.unwrap();

        let msg = Message::new(b"hello".to_vec(), "src");
        pub_.publish(&msg, "dest").await.unwrap();

        assert_eq!(pub_.len(), 1);
        let stored = pub_.messages();
        assert_eq!(stored[0].payload, b"hello");
        assert_eq!(stored[0].topic, "dest");
    }

    #[tokio::test]
    async fn mock_source_generates_messages() {
        let mut src = MockSource::new().with_interval(Duration::from_millis(1));
        src.connect().await.unwrap();
        src.subscribe(&["test-topic"]).await.unwrap();

        let msg = src.consume().await.unwrap().unwrap();
        assert_eq!(msg.topic, "test-topic");
        assert!(!msg.payload.is_empty());
        assert_eq!(msg.get_header("source", ""), "mock");
    }
}
